<a id="item31"></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np


<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download the data and read it into a <em>pandas</em> dataframe.

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

### Normalization of data

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
print(n_cols)

8


### Split the data into training and testing sets

let's import `train_test_split` from the `sklearn`

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)

<a id="item1"></a>

<a id='item32'></a>

## Import Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

#### Let's go ahead and import the Keras library

In [12]:
import keras

Using TensorFlow backend.


As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [13]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has three hidden layers, each of 10 hidden units.

<a id="item4"></a>

<a id='item34'></a>

## Train and Test the Network

Let's call the function now to create our model.

In [15]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.

In [16]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 1s - loss: 1668.1176 - val_loss: 1195.0188
Epoch 2/50
 - 0s - loss: 1636.1871 - val_loss: 1159.8586
Epoch 3/50
 - 0s - loss: 1584.3494 - val_loss: 1099.8286
Epoch 4/50
 - 0s - loss: 1499.3044 - val_loss: 1007.0664
Epoch 5/50
 - 0s - loss: 1369.5083 - val_loss: 866.6372
Epoch 6/50
 - 0s - loss: 1176.4759 - val_loss: 684.0180
Epoch 7/50
 - 0s - loss: 938.1720 - val_loss: 488.6310
Epoch 8/50
 - 0s - loss: 690.8218 - val_loss: 328.6202
Epoch 9/50
 - 0s - loss: 494.1730 - val_loss: 240.3371
Epoch 10/50
 - 0s - loss: 383.5754 - val_loss: 204.4718
Epoch 11/50
 - 0s - loss: 331.5479 - val_loss: 190.2490
Epoch 12/50
 - 0s - loss: 298.7475 - val_loss: 185.9808
Epoch 13/50
 - 0s - loss: 275.9404 - val_loss: 179.3193
Epoch 14/50
 - 0s - loss: 259.0850 - val_loss: 176.3881
Epoch 15/50
 - 0s - loss: 244.8036 - val_loss: 171.5628
Epoch 16/50
 - 0s - loss: 233.7600 - val_loss: 168.8559
Epoch 17/50
 - 0s - loss: 225.3934 - val_loss: 165.4893
E

In [17]:
model.evaluate(X_test,y_test, verbose=1)

309/309 [==============================] - 0s 36us/step


115.33065494673151

   Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

   Let's import the mean_squared_error function from Scikit-learn.

In [18]:
from sklearn.metrics import mean_squared_error

In [19]:
y_pred = model.predict(X_test)

In [20]:
y_pred

array([[54.90538  ],
       [49.014217 ],
       [67.30542  ],
       [48.580784 ],
       [31.246977 ],
       [35.47047  ],
       [24.062439 ],
       [50.31472  ],
       [29.834429 ],
       [31.946722 ],
       [23.09238  ],
       [23.402786 ],
       [69.096535 ],
       [48.75457  ],
       [30.38541  ],
       [39.19153  ],
       [29.809925 ],
       [23.869684 ],
       [27.530308 ],
       [29.09488  ],
       [31.376404 ],
       [35.737045 ],
       [34.43791  ],
       [24.59634  ],
       [31.745306 ],
       [33.46974  ],
       [16.64486  ],
       [39.182674 ],
       [48.342976 ],
       [21.299839 ],
       [39.45981  ],
       [28.250652 ],
       [41.22976  ],
       [54.10766  ],
       [27.492577 ],
       [35.566097 ],
       [29.386143 ],
       [32.414658 ],
       [19.891844 ],
       [36.92704  ],
       [20.123266 ],
       [20.381027 ],
       [24.175024 ],
       [47.26065  ],
       [21.947807 ],
       [52.854725 ],
       [52.525394 ],
       [55.38

In [21]:
mse = mean_squared_error(y_test, y_pred)
print('mse: ', mse)

mse:  115.33065696564404


Let's create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [22]:
n = 50
epochs = 50
mean_squared_errors = []
for i in range(0, n):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    mse = model.evaluate(X_test, y_test, verbose=0)
    print("mse " +str(i+1)+": ", mse)
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)
    
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)   

print("Mean : ", mean)
print("Standard Deviation : ", standard_deviation)

mse 1:  73.5371558673945
mse 2:  52.414789884993176
mse 3:  50.335600349895394
mse 4:  38.27787804217786
mse 5:  43.630825808133125
mse 6:  40.128428314110224
mse 7:  40.88536527318862
mse 8:  35.51020978415282
mse 9:  36.04054683697648
mse 10:  31.621793802502086
mse 11:  33.6104108507965
mse 12:  33.27687330770647
mse 13:  35.06964666866562
mse 14:  39.7074046860235
mse 15:  34.65724141389421
mse 16:  26.10369811320382
mse 17:  29.559300660315454
mse 18:  33.48379878936076
mse 19:  28.047822186861996
mse 20:  31.738841917908307
mse 21:  35.43041264432148
mse 22:  30.27265977087916
mse 23:  29.024348842287527
mse 24:  34.56836822040644
mse 25:  26.024196458094327
mse 26:  29.00008157156046
mse 27:  35.255320589133454
mse 28:  27.347859256090082
mse 29:  32.06199912654544
mse 30:  25.175706635015295
mse 31:  27.0077102006832
mse 32:  31.474947339894317
mse 33:  28.987984703582466
mse 34:  26.355035318911654
mse 35:  31.59087053317468
mse 36:  24.586560795608076
mse 37:  27.760332450125

In [23]:
print('Below are the mean and standard deviation of 50 mean squared errors of normalized data for 50 epochs')
print("Mean : ", mean)
print("Standard Deviation : ", standard_deviation)

Below are the mean and standard deviation of 50 mean squared errors of normalized data for 50 epochs
Mean :  32.792847370116995
Standard Deviation :  8.337224925941959


In [24]:
print('mean squared errors: ', mean_squared_errors)

mean squared errors:  [73.53715307 52.41478884 50.33560044 38.27787585 43.63082578 40.12842777
 40.8853652  35.51020886 36.04054662 31.62179467 33.61041155 33.27687326
 35.06964647 39.70740447 34.65724181 26.10369817 29.55930021 33.48379743
 28.04782109 31.73884167 35.43041335 30.2726591  29.02435046 34.56836876
 26.024197   29.00008141 35.25531972 27.34785834 32.06199821 25.17570633
 27.00770903 31.47494786 28.98798308 26.35503334 31.59086932 24.58655989
 27.76033246 30.00101984 29.42904888 26.51398168 30.16214728 26.98093639
 29.40028322 34.84454746 24.07200343 26.42308395 25.91297253 27.85192072
 29.91196857 28.57644369]


The `mean` of 50 mean squared errors in part D is 32 while it was 51 in part B for 50 epochs.